File: src/com/kingmang/lazurite/parser/AST/Expressions/ObjectCreationExpression.java
package com.kingmang.lazurite.parser.AST.Expressions;

import com.kingmang.lazurite.exceptions.LZRException;
import com.kingmang.lazurite.core.*;
import com.kingmang.lazurite.parser.AST.Statements.ClassDeclarationStatement;
import com.kingmang.lazurite.parser.AST.Statements.FunctionDefineStatement;
import com.kingmang.lazurite.patterns.visitor.ResultVisitor;
import com.kingmang.lazurite.patterns.visitor.Visitor;
import com.kingmang.lazurite.runtime.ClassInstanceValue;
import com.kingmang.lazurite.runtime.ClassMethod;
import com.kingmang.lazurite.runtime.Value;
import com.kingmang.lazurite.runtime.Variables;
import lombok.AllArgsConstructor;
import lombok.Getter;


import java.util.Iterator;
import java.util.List;

@AllArgsConstructor
public final class ObjectCreationExpression implements Expression {
    
    public final String className;
    public final List<Expression> constructorArguments;
    
    @Override
    public Value eval() {
        final ClassDeclarationStatement cd = ClassDeclarations.get(className);
        if (cd == null) {
            // Is Instantiable?
            if (Variables.isExists(className)) {
                final Value variable = Variables.get(className);
                if (variable instanceof Instantiable) {
                    return ((Instantiable) variable).newInstance(ctorArgs());
                }
            }
            throw new LZRException("UnknownClassException ","Unknown class " + className);
        }
        

        final ClassInstanceValue instance = new ClassInstanceValue(className);
        for (AssignmentExpression f : cd.fields) {

            final String fieldName = ((VariableExpression) f.target).name;
            instance.addField(fieldName, f.eval());
        }
        for (FunctionDefineStatement m : cd.methods) {
            instance.addMethod(m.name, new ClassMethod(m.arguments, m.body, instance));
        }
        

        instance.callConstructor(ctorArgs());
        return instance;
    }
    
    private Value[] ctorArgs() {
        final int argsSize = constructorArguments.size();
        final Value[] ctorArgs = new Value[argsSize];
        for (int i = 0; i < argsSize; i++) {
            ctorArgs[i] = constructorArguments.get(i).eval();
        }
        return ctorArgs;
    }
    
    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }

    @Override
    public <R, T> R accept(ResultVisitor<R, T> visitor, T t) {
        return visitor.visit(this, t);
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder();
        sb.append("new ").append(className).append(' ');
        final Iterator<Expression> it = constructorArguments.iterator();
        if (it.hasNext()) {
            sb.append(it.next());
            while (it.hasNext()) {
                sb.append(", ").append(it.next());
            }
        }
        sb.append(')');
        return sb.toString();
    }
}

File: src/main/java/com/kingmang/lazurite/parser/ast/expressions/ObjectCreationExpression.kt
package com.kingmang.lazurite.parser.ast.expressions

import com.kingmang.lazurite.core.ClassDeclarations
import com.kingmang.lazurite.core.Instantiable
import com.kingmang.lazurite.exceptions.LzrException
import com.kingmang.lazurite.patterns.visitor.ResultVisitor
import com.kingmang.lazurite.patterns.visitor.Visitor
import com.kingmang.lazurite.runtime.ClassInstanceBuilder
import com.kingmang.lazurite.runtime.Variables.get
import com.kingmang.lazurite.runtime.Variables.isExists
import com.kingmang.lazurite.runtime.values.LzrValue

data class ObjectCreationExpression(
    val className: String,
    val constructorArguments: List<Expression>
) : Expression {
    override fun eval(): LzrValue {
        val cd = ClassDeclarations[className]
        if (cd == null) {
            // Is Instantiable?
            if (isExists(className)) {
                val variable = get(className)
                if (variable is Instantiable) {
                    return variable.newInstance(ctorArgs())
                }
            }
            throw LzrException("UnknownClassException", "Unknown class $className")
        }


        val builder = ClassInstanceBuilder(className)
        for (f in cd.fields) {
            val fieldName = (f.target as VariableExpression).name
            builder.addField(fieldName, f.eval())
        }
        for (m in cd.methods) {
            builder.addMethod(m.name, m.arguments, m.body)
        }
        return builder.build(ctorArgs())
    }

    private fun ctorArgs(): Array<LzrValue> =
        constructorArguments.map(Expression::eval).toTypedArray()

    override fun accept(visitor: Visitor) =
        visitor.visit(this)

    override fun <R, T> accept(visitor: ResultVisitor<R, T>, input: T): R? =
        visitor.visit(this, input)

    override fun toString(): String {
        val sb = StringBuilder()
        sb.append("new ").append(className).append(' ')
        val it = constructorArguments.iterator()
        if (it.hasNext()) {
            sb.append(it.next())
            while (it.hasNext()) {
                sb.append(", ").append(it.next())
            }
        }
        sb.append(')')
        return sb.toString()
    }
}
